engine/collection/epochmgr/epoch_components.go

```diff
@@ -0,0 +1,107 @@
+package epochmgr
+
+import (
+	"context"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/util"
+	"github.com/onflow/flow-go/state/cluster"
+)
+
+// EpochComponents represents all dependencies for running an epoch.
+type EpochComponents struct {
+	*component.ComponentManager
+	state             cluster.State
+	prop              component.Component
+	sync              module.ReadyDoneAware
+	hotstuff          module.HotStuff
+	voteAggregator    hotstuff.VoteAggregator
+	timeoutAggregator hotstuff.TimeoutAggregator
+}
+
+var _ component.Component = (*EpochComponents)(nil)
+
+func NewEpochComponents(
+	state cluster.State,
+	prop component.Component,
+	sync module.ReadyDoneAware,
+	hotstuff module.HotStuff,
+	voteAggregator hotstuff.VoteAggregator,
+	timeoutAggregator hotstuff.TimeoutAggregator,
+) *EpochComponents {
+	components := &EpochComponents{
+		state:             state,
+		prop:              prop,
+		sync:              sync,
+		hotstuff:          hotstuff,
+		voteAggregator:    voteAggregator,
+		timeoutAggregator: timeoutAggregator,
+	}
+
+	builder := component.NewComponentManagerBuilder()
+	// start new worker that will start child components and wait for them to finish
+	builder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		// start vote and timeout aggregators, hotstuff will be started by compliance engine
+		voteAggregator.Start(ctx)
+		timeoutAggregator.Start(ctx)
+		// wait until all components start
+		<-util.AllReady(components.prop, components.sync, components.voteAggregator, components.timeoutAggregator)
+
+		// signal that startup has finished, and we are ready to go
+		ready()
+
+		// wait for shutdown to be commenced
+		<-ctx.Done()
+		// wait for compliance engine and event loop to shut down
+		<-util.AllDone(components.prop, components.sync, components.voteAggregator, components.timeoutAggregator)
```

Comment:
can itself shut down before other components to shut down?

would it miss a few events if it's shut down earlier?

Reply:
> can itself shut down before other components to shut down?

Do you mean, "can the `epochmgr` engine shut down before the epoch components shut down?" (What is `itself` referring to?)

> would it miss a few events if it's shut down earlier?

Yes, although I think we can miss events regardless of the order in which we are shutting down. Shutting down means, we are either turning off an old cluster consensus or shutting down the whole node. In neither case are missed events problematic from my perspective.

Reply:
Oh, I see. the ctx is actually the shut down signal itself. 

I thought it was an engine's Done channel, actually we are moving away from that pattern.